packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1292 @@
+  ///    even if location does not change.
+  static void neglect(BuildContext context, VoidCallback callback) {
+    final _RouterScope scope = context
+      .getElementForInheritedWidgetOfExactType<_RouterScope>()
+      .widget as _RouterScope;
+    scope.routerState._setStateWithExplicitReportStatus(_IntentionToReportRouteInformation.ignore, callback);
+  }
+
+  @override
+  StatefulElement createElement() => _RouterStatefulElement(this);
+
+  @override
+  State<Router<T>> createState() => _RouterState<T>();
+}
+
+typedef _AsyncPassthrough<Q> = Future<Q> Function(Q);
+
+class _RouterStatefulElement extends StatefulElement {
+  _RouterStatefulElement(StatefulWidget widget) : super(widget);
+
+  _RouterState<dynamic> get routerState => state as _RouterState<dynamic>;
+
+  @override
+  void rebuild() {
+    super.rebuild();
+    routerState._reportRouteInformation();
+  }
+}
+
+// Whether to report the route information in this build cycle.
+enum _IntentionToReportRouteInformation {
+  // We haven't receive any signal on whether to report.
+  none,
+  // Report if route information changes.
+  maybe,
+  // Report regardless of route information changes.
+  must,
+  // Don't report regardless of route information changes.
+  ignore,
+}
+
+class _RouterState<T> extends State<Router<T>> {
+  Object _currentRouteInformationParserTransaction;
+  Object _currentRouterDelegateTransaction;
+  _IntentionToReportRouteInformation _currentIntentionToReporting;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.routeInformationProvider
+      ?.addListener(_handleRouteInformationProviderNotification);
+    widget?.backButtonDispatcher
+      ?.addCallback(_handleBackButtonDispatcherNotification);
+    widget.routerDelegate.addListener(_handleRouterDelegateNotification);
+    _currentIntentionToReporting = _IntentionToReportRouteInformation.none;
+    if (widget.routeInformationProvider != null) {
+      _processInitialRoute();
+    }
+    _lastReportedLocation = widget.routeInformationProvider?.value?.location;
+  }
+
+  String _lastReportedLocation;
+  void _reportRouteInformation() {
+    switch (_currentIntentionToReporting) {
+      case _IntentionToReportRouteInformation.none:
+        _currentIntentionToReporting = _IntentionToReportRouteInformation.none;
+        return;
+      case _IntentionToReportRouteInformation.ignore:
+        // In the ignore case, we still want to update _lastReportedRouteName.
+      case _IntentionToReportRouteInformation.maybe:
+      case _IntentionToReportRouteInformation.must:
+        final T configuration = widget.routerDelegate.currentConfiguration;
+        if (configuration != null) {
+          final RouteInformation routeInformation = widget
+            .routeInformationParser
+            .restoreRouteInformation(configuration);
+          assert(routeInformation != null);
+          if (_currentIntentionToReporting != _IntentionToReportRouteInformation.ignore &&
```

Comment:
use switch/case with enums.

Instead of lumping all cases into one in this switch/case, maybe separate them out. It would also make the code easier to read. You can factor common logic out into private methods.